src/main/java/com/carrotcreative/recyclercore/widget/ProgressRecyclerViewLayout.java

```diff
@@ -117,4 +124,131 @@ public void stopScroll()
     {
         mRecyclerView.stopScroll();
     }
+
+    // ========================================== //
+    // =============  Error State =============== //
+    // ========================================== //
+
+    public boolean isShowingErrorState()
+    {
+        if(mErrorStateView != null && mErrorStateView.getVisibility() == View.VISIBLE)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    public void setErrorStateEnabled(boolean show)
+    {
+        if(mErrorStateView == null)
+        {
+            throw new IllegalStateException("Trying to setErrorStateEnabled without setting the error state View.");
+        }
+
+        if(show)
+        {
+            if(! isShowingErrorState())
+            {
+                saveCurrentViewState();
+                mErrorStateView.setVisibility(VISIBLE);
+                mRecyclerView.setVisibility(GONE);
+                mProgressBar.setVisibility(GONE);
+            }
+        }
+        else
+        {
+            restorePreviousViewState();
+        }
+    }
+
+    public void setErrorStateEnabled(View errorView, boolean show)
+    {
+        if(mErrorStateView == null)
+        {
+            mErrorStateView = errorView;
+            mContainer.addView(mErrorStateView);
+            mErrorStateView.setVisibility(GONE);
+        }
+
+        if(mErrorStateView != errorView)
+        {
+            mContainer.removeView(mErrorStateView);
+            mErrorStateView = errorView;
+            mContainer.addView(mErrorStateView);
+            mErrorStateView.setVisibility(GONE);
+        }
+
+        setErrorStateEnabled(show);
+    }
+
+    private void saveCurrentViewState()
+    {
+        if(mPrevViewVisibilityInstanceState == null)
+        {
+            mPrevViewVisibilityInstanceState = new ViewVisibilityInstanceState();
+        }
+
+        mPrevViewVisibilityInstanceState.setProgressViewVisibility(mProgressBar.getVisibility());
+        mPrevViewVisibilityInstanceState.setRecyclerViewVisibility(mRecyclerView.getVisibility());
+        mPrevViewVisibilityInstanceState.setErrorViewVisibility(mErrorStateView.getVisibility());
+    }
+
+    private void restorePreviousViewState()
+    {
+        if(mPrevViewVisibilityInstanceState != null)
+        {
+            mRecyclerView.setVisibility(mPrevViewVisibilityInstanceState.getRecyclerViewVisibility());
+            mProgressBar.setVisibility(mPrevViewVisibilityInstanceState.getProgressViewVisibility());
+            mErrorStateView.setVisibility(mPrevViewVisibilityInstanceState.getErrorViewVisibility());
+        }
+    }
+
+    /**
+     * Add Visibility annotation for lint checks.
+     */
+    @IntDef({VISIBLE, INVISIBLE, GONE})
```

Comment:
So, I found this piece in the Android's View.java class, 
```
    @IntDef({VISIBLE, INVISIBLE, GONE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Visibility {}
```

```
@Visibility
    public int getVisibility() {
        return mViewFlags & VISIBILITY_MASK;
    }
```

They have added a check that the passed the checks so we like try to pass integer type 0, 8 for the visibility change, it  works as a runtime but shows a warning during compile time. 
So I added this to counter that. :smile: